packages/x-data-grid-premium/src/hooks/features/dataSource/useGridDataSourcePremium.tsx

```diff
@@ -0,0 +1,86 @@
+import * as React from 'react';
+import {
+  useGridApiEventHandler as addEventHandler,
+  useGridApiMethod,
+  useGridSelector,
+  GridEventLookup,
+  GRID_ROOT_GROUP_ID,
+  GridValidRowModel,
+} from '@mui/x-data-grid-pro';
+import {
+  useGridDataSourceBase,
+  GridStrategyProcessor,
+  useGridRegisterStrategyProcessor,
+  GridStrategyProcessorName,
+} from '@mui/x-data-grid-pro/internals';
+import { GridPrivateApiPremium } from '../../../models/gridApiPremium';
+import { DataGridPremiumProcessedProps } from '../../../models/dataGridPremiumProps';
+import { GridDataSourcePremiumPrivateApi, GridGetRowsResponsePremium } from './models';
+import { gridAggregationModelSelector } from '../aggregation/gridAggregationSelectors';
+
+export const useGridDataSourcePremium = (
+  apiRef: React.MutableRefObject<GridPrivateApiPremium>,
+  props: DataGridPremiumProcessedProps,
+) => {
+  const { api, strategyProcessor, events } = useGridDataSourceBase<GridPrivateApiPremium>(
+    apiRef,
+    props,
+  );
+  const aggregationModel = useGridSelector(apiRef, gridAggregationModelSelector);
+  const aggregateRowRef = React.useRef<GridValidRowModel>({});
+
+  const handleDataUpdate = React.useCallback<GridStrategyProcessor<'dataSourceRowsUpdate'>>(
```

Comment:
Do you mean something like publishing an event like we initially did in the lazy loading PR?

Again no hard preference but I wanted to have all the code of a function co-located.
1. We know for a specific version what exactly happens with the response data at a single place.
2. We only duplicate the code of those methods which are different. 
3. It will be easier to extend the code, for example when porting code related to row grouping to the premium package.
4. (in the future) we could even have different code paths for the same functions based on the package.
 
These points are arguable though. Wdyt?

Reply:
>Do you mean something like publishing an event like we initially did in the lazy loading PR

No, I meant to have the base a bit more closed down.
I think that we have a different expectation of `base` :)

You use it more as an initial (base) hook for which you override part that you want to function differently.
For me it would be more a hook which API has parameters that are used in specific code branches inside base.

That would correspond to the strategy approach where we have
```
const getRowsResponse = await getRows(fetchParams);

const cacheResponses = cacheChunkManager.splitResponse(fetchParams, getRowsResponse);
cacheResponses.forEach((response, key) => {
  cache.set(key, response);
});

if (lastRequestId.current === requestId) {
  apiRef.current.applyStrategyProcessor('dataSourceRowsUpdate', {
    response: getRowsResponse,
    fetchParams,
  });
}
```
data source always sets the cache, you can process rows.

Here it would be, base always sets rows and row count, because that always happens
specific hooks can continue with something else (premium adding aggregation processing)

What I am trying to prevent is having to many different patterns for dealing with the differences between packages.

Reply:
In addition to this, I think that the aggregation piece should be separate (it can also be a follow up), because you need the same approach with lazy loading
Have strategy do its things and then aggregation add on top of that, right?

Reply:
> Here it would be, base always sets rows and row count, because that always happens
specific hooks can continue with something else (premium adding aggregation processing)

I feel like piped processing within strategy processing could be a nice long term approach for this.

The handler in the base hook would become:

```tsx
const handleDataUpdate = React.useCallback<GridStrategyProcessor<'dataSourceRowsUpdate'>>(
  (params) => {
    const { response } = params;
    if (response.rowCount !== undefined) {
      apiRef.current.setRowCount(response.rowCount);
    }
    apiRef.current.setRows(response.rows);
	apiRef.current.unstable_applyPipeProcessors('dataSourceRowsUpdate', { params, response })
  },
  [apiRef],
);
```

The pipe processor in the aggregation would continue from here:

```tsx
const processDataSourceRows = React.useCallback<GridPipeProcessor<'dataSourceRowsUpdate'>>(
  ({ params, response }) => {
    if (response.aggregateRow) {
      aggregateRowRef.current = response.aggregateRow;
    }
    if (Object.keys(params.aggregationModel).length > 0) {
      apiRef.current.requestPipeProcessorsApplication('hydrateRows');
      apiRef.current.applyAggregation();
    }
	// return the same arguments for the next plugin to process
    return {
      params,
      response,
    };
  },
  [apiRef],
);
```

Similarly, the other plugins can provide a pipe processor for their specific use-case. Is it aligned with your expectations?

> In addition to this, I think that the aggregation piece should be separate (it can also be a follow up), because you need the same approach with lazy loading
> Have strategy do its things and then aggregation add on top of that, right?

I'm not sure about this, the aggregation use-case sounds like a synchronous behavior rather than asynchronous. As soon as the new rows are received, the aggregation should be applied to avoid any lags.

What do you think about the lazy-loading? Would it be better off delaying the processing on it?